src/psyclone/psyir/tools/dependency_tools.py

```diff
@@ -165,59 +243,15 @@ def array_access_parallelisable(self, loop_variable, var_info):
-        # in which dimension for the component. For example, `a%b%c(i,j)`
-        # would store (2,1) for an access to j - component 2 (which is c),
-        # and 2nd dimension (j).
-        # TODO #1269: code duplicated with nemo_loop_fuse
-        found_dimension_index = None
-
         # Additionally, collect all indices that are actually used, since
         # they are needed in a test further down.
-        all_indices = []
-
-        # Loop over all the accesses of this variable
-        for access in var_info.all_accesses:
-            component_indices = access.component_indices
 
-            # Now determine all dimensions that depend
-            # on the parallel variable. This outer loop is over
-            # the indices used for each component, e.g.
-            # a(i,j)%b(k) it would first handle `(i,j)`, then
-            # `(k)`.
-            for component_index, index_expressions in \
-                    enumerate(component_indices):
-
-                # This inner loop loop over all indices for the
-                # current component, i.e. `[i, j]` for the first
-                # component above, then `[k]`.
-                for dimension_index, index_expression in \
-                        enumerate(index_expressions):
-                    # Add all the variables used in the index expression
-                    accesses = VariablesAccessInfo()
-                    index_expression.reference_accesses(accesses)
-                    if Signature(loop_variable) not in accesses:
-                        continue
-
-                    # if a previously identified index location does not match
-                    # the current index location (e.g. a(i,j), and a(j,i) ),
-                    # then the loop can not be parallelised
-                    ind_pair = (component_index, dimension_index)
-                    if found_dimension_index and \
-                            found_dimension_index != ind_pair:
-                        # TODO #1268: improve error message
-                        self._add_warning("Variable '{0}' is using loop "
-                                          "variable '{1}' in index '{2}'' "
-                                          "and '{3}'."
-                                          .format(var_info.var_name,
-                                                  loop_variable,
-                                                  found_dimension_index,
-                                                  ind_pair))
-                        return False
-                    found_dimension_index = ind_pair
-                    all_indices.append(index_expression)
+        # Detect if this variable adds a new message, and if so, abort early
+        old_num_messages = len(self.get_all_messages())
+        all_indices = \
+            self.array_accesses_consistent(signature, loop_variable, var_info)
+        if len(self.get_all_messages()) > old_num_messages:
```

Comment:
Not something for this PR but it feels like the dependency checking should be configurable - i.e. it would be good if we could set the 'log level' to abort with an exception on first error. We could then use try..except which is better than counting error messages.

Reply:
Yes, that code is particularly ugly :( My idea that the dependency tool leaves it to the caller to decide how to handle error (and especially allowing the dependency tools to collect as many errors as possible, so that the user can look at them all at once and fix them at once) breaks down the moment we have a function that is used internally and externally :( To you want me to open a ticket for that?

Reply:
Yes please. Maybe changing it to do logging might help? Also, as I say earlier, I think the `array_accesses_consistent` method should really wrap some of this and just return True or False.

Reply:
Fixed when refactoring `array_access_consistent`.